### ISSUE TITLE
pre-diff hooks

### ISSUE THREAD

**Issue by ansuz** [fiduswriter/diffDOM#36]
I'm using diffDOM for applying patches to a contentEditable section of a page, and it's working really well. Unfortunately, patching a DOM while you're editing has unwanted side effects that often displace the cursor selection.

I'm currently modifying parts of the diff algorithm to execute optional hooks that would expose information about the nodes which are being patched, which would make it much easier to gracefully handle cursor functionality. In general, I'd like to extend the library to address issues which are pertinent to diff/patching a contentEditable node.

I'd prefer to get these changes pushed upstream so I don't have to maintain a fork. Is there an IRC channel I can join to discuss changes more directly, or should I just submit a PR and go from there?

PS: Thanks for the excellent library, I tried a number of alternatives before settling on diffDOM.

**Comment by johanneswilm** [comment]
Hey, thanks!

It's interesting you would mention contenteditable -- that's actually what we used it for for a long time. The main reason we switched to using a framework (ProseMirror) was that maintaining an entire editing framework ourselves ended up being too much work.

What this means is essentially two things that should concern you:

1. You probably noticed already that it was made with text editing in mind. For example, the ability to hook in google's diff-match-patch for  text node differences. Also, we allow sibling text nodes, which is something you may end up with if combining JavaScript and contenteditable.

2. The library currently has no specific purpose for us. I will continue to maintain it, but we don't have set priorities for what goes in or doesn't. Basically anything that makes it usable for a greater number of users, while not slowing it down (considerably) or making it unusable for existing users is something I will be OK with merging. So unless there is something specific you feel may be controversial that you need to discuss, I would say a PR is enough. If you add tests for features you add, your patches will be even more welcome.

**Comment by ansuz** [comment]
Wonderful!

So far I've just added hooks to be executed before the switch statements in `applyVirtualDiff` and `applyDiff`. The actual logic for figuring out what to do with the information exposed by those hooks would be written into my application, not the core.

Aside from the trivial overhead of executing an empty function there shouldn't be any significant changes. I benchmarked against guarding a function call with if statements and logical-and expressions, and the empty function call performed the best, by a factor of 2-1.

One of my requirements is that my code has to extend [CKEditor](http://ckeditor.com/), so an all-in-one solution isn't really an option.

I'll keep tinkering until I'm done making changes, then submit a PR.

Thanks for your help!
